File: RetroGameGauntlet.Forms/Services/FlickrImageSearchService.cs
using System.Linq;
using System.Threading.Tasks;
using System.Net.Http;
using System.Net;
using Newtonsoft.Json;
using System.Diagnostics;
using RetroGameGauntlet.Forms.Models.Api;

namespace RetroGameGauntlet.Forms.Services
{
    public class FlickrImageSearchService
    {
        private const string ApiKey = "78d991c72ad7a124dcef3c2796d6f4ea";

        private async Task<string> GetImageFromLink(string link)
        {
            HttpResponseMessage response;
            try 
            {
                response = await new HttpClient().GetAsync(link);
            }
            catch (TaskCanceledException)
            {
                return null;
            }
            catch (WebException)
            {
                return null;
            }
            if (response.StatusCode != HttpStatusCode.OK) 
            {
                return null;
            }

            var responseBody = response.Content.ReadAsStringAsync().Result;
            Debug.WriteLine(string.Format("Request URL is {0}, response JSON is {1}", link, responseBody));
            var responseObject = JsonConvert.DeserializeObject<FlickApiSearchResponseModel>(responseBody);

            if (responseObject?.Photos?.Photo?.Any() == true)
            {
                var photo = responseObject?.Photos?.Photo?.First();
                if (photo != null)
                {
                    return "https://" + "farm" + photo.Farm + ".staticflickr.com/" + photo.Server + "/" + photo.Id + "_" + photo.Secret + "_z" + ".jpg";
                }
            }
            return null;
        }

        public async Task<string> GetImage()
        {
            return await GetImageFromLink(GetLinkForQuery("retro game"));
        }

        public async Task<string> GetImageForGame(string gameName, string platformName)
        {
            var img = await GetImageFromLink(GetLinkForQuery(gameName + " " + platformName));
            return img ?? await GetImageFromLink(GetLinkForQuery(platformName));
        } 

        private string GetLinkForQuery(string query) 
        {
            return "https://api.flickr.com/services/rest/"
                            + "?format=" + "json"
                            + "&nojsoncallback=" + "1"
                            + "&method=" + "flickr.photos.search"
                            + "&api_key=" + ApiKey
                            + "&per_page=" + "1"
                            + "&text=" + query;
        }
    }
}

File: RetroGameGauntlet.PCL/Services/FlickrImageSearchService.fs
namespace RetroGameGauntlet.PCL.Services

open Newtonsoft.Json
open RetroGameGauntlet.PCL
open RetroGameGauntlet.PCL.Models.API
open RetroGameGauntlet.PCL.Services
open System.Diagnostics
open System.Linq
open System.Net.Http
open System.Net

type FlickrImageSearchService() = 
    let apiKey: string = "78d991c72ad7a124dcef3c2796d6f4ea";

    let getLinkForQuery query:string =
        "https://api.flickr.com/services/rest/" 
        + "?format=" + "json"
        + "&nojsoncallback=" + "1"
        + "&method=" + "flickr.photos.search"
        + "&api_key=" + apiKey
        + "&per_page=" + "1"
        + "&text=" + query

    interface IImageSearchService with
        member this.GetImage =
            async {
                let link = this.GetLinkForQuery("retro game")
                let! a = this.GetImageFromLink(link)
                return if a.IsSome then a.Value else System.String.Empty
            }
            |> Async.StartAsTask

        member this.GetImageForGame gameName platformName = 
            async {
                let! img = this.GetImageFromLink(this.GetLinkForQuery(gameName + " " + platformName))
                return if img.IsSome then img.Value else System.String.Empty
            }
            |> Async.StartAsTask


    member private this.GetImageFromLink (link: string) : Async<Option<string>> = 
        async {
            try
                Debug.WriteLine("GetImageFromLink " + link)
                let! response = RetroGameGauntletCore.HttpClient.GetAsync(link) |> Async.AwaitTask
                let! responseBody = response.Content.ReadAsStringAsync() |> Async.AwaitTask
                Debug.WriteLine(System.String.Format("Request URI is {0}, response JSON is {1}", link, responseBody))
                let responseObject = JsonConvert.DeserializeObject<FlickSearchResponseModel>(responseBody)
                if responseObject.Photos.Photo.Any() then
                    let photo = responseObject.Photos.Photo.First()
                    let rvalue = System.String.Format("https://farm{0}.staticflickr.com/{1}/{2}_{3}_z.jpg",
                                                      photo.Farm,
                                                      photo.Server,
                                                      photo.Id,
                                                      photo.Secret)
                    return Some rvalue
                else
                    return None
            with
            | e ->
                Debug.WriteLine("Error in GetImageFromLink" + e.ToString())
                return None
        }

    member private this.GetLinkForQuery (query: string) =
        "https://api.flickr.com/services/rest/"
                        + "?format=" + "json"
                        + "&nojsoncallback=" + "1"
                        + "&method=" + "flickr.photos.search"
                        + "&api_key=" + apiKey
                        + "&per_page=" + "1"
                        + "&text=" + query;
